### PR TITLE
Add spacing to regionprops and moments.

### PR DESCRIPTION
**Pull request by tibuch**
## Description
Enhancement, closes https://github.com/scikit-image/scikit-image/issues/5112.

## Checklist

- [x] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [ ] Gallery example in `./doc/examples` (new features only)
- [x] Changes are covered by an  existing benchmark
- [x] Unit tests
- [ ] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- [x] Descriptive commit messages (see below)

## For reviewers

- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
  `doc/release/release_dev.rst`.
- There is a bot to help automate backporting a PR to an older branch. For
  example, to backport to v0.19.x after merging, add the following in a PR
  comment: `@meeseeksdev backport to v0.19.x`
- To run benchmarks on a PR, add the `run-benchmark` label. To rerun, the label
  can be removed and then added again. The benchmark output can be checked in
  the "Actions" tab.

### CODE REVIEW

**Comment by grlee77** [comment]
Thanks @tibuch, we are very keen to have this functionality. I will try to review this soon

File: skimage/measure/_regionprops.py [hunk 1]
```diff
@@ -397,8 +405,8 @@ def image_convex(self):
     @property
     def coords(self):
         indices = np.nonzero(self.image)
-        return np.vstack([indices[i] + self.slice[i].start
-                          for i in range(self._ndim)]).T
+        return np.vstack([(indices[i] + self.slice[i].start) * s
+                          for i, s in zip(range(self._ndim), self._spacing)]).T
```
Comment (jni) [inline skimage/measure/_regionprops.py:409]: imho these coordinates are for indexing into the image and should not be scaled.

Perhaps we should have `coords` and `coords_scaled`, or `indices` and `coords`? And the docstring should state which property is scaled and which is not...

File: skimage/measure/_regionprops.py [hunk 2]
```diff
@@ -359,10 +362,15 @@ def __setattr__(self, name, value):
         else:
             super().__setattr__(name, value)
 
+    @property
+    @_cached
+    def pixel_area(self):
```
[the remark this answers is not in the document]
Comment (jni) [reply skimage/measure/_regionprops.py:367]: Additionally, I do think many people will want to know the area in pixels/voxels, so maybe we can now add a property like `num_pixels`?

File: skimage/measure/_regionprops.py [hunk 3]
```diff
@@ -530,19 +538,21 @@ def axis_minor_length(self):
...
         mu = _moments.moments_central(self.image.astype(np.uint8),
-                                      self.centroid_local, order=3)
+                                      self.centroid_local, order=3, spacing=self._spacing)
         return mu
 
     @property
     @only2d
     def moments_hu(self):
+        if not (np.array(self._spacing) == np.array([1, 1])).all():
+            raise NotImplementedError('`moments_hu` supports spacing = (1, 1) only')
```
Comment (jni) [inline skimage/measure/_regionprops.py:555]: Ha, I wish I'd thought of this when I was trying to implement this :rofl: :+1:

File: skimage/measure/_regionprops.py [hunk 4]
```diff
@@ -1253,14 +1274,14 @@ def regionprops(label_image, intensity_image=None, cache=True,
...
-                    'the True values as a single label.')
+                'Non-integer image types are ambiguous: '
+                'use skimage.measure.label to label the connected'
+                'components of label_image,'
+                'or label_image.astype(np.uint8) to interpret'
+                'the True values as a single label.')
         else:
             raise TypeError(
-                    'Non-integer label_image types are ambiguous')
+                'Non-integer label_image types are ambiguous')
```
Comment (jni) [inline skimage/measure/_regionprops.py:1284]: Do you mind reverting unrelated formatting changes? They will make git-blame harder to follow.

File: skimage/measure/tests/test_regionprops.py [hunk 5]
```diff
@@ -198,51 +338,75 @@ def test_coordinates():
     prop_coords = regionprops(sample)[0].coords
     assert_array_equal(prop_coords, coords)
 
+    spacing = (1, 0.5)
+    prop_coords = regionprops(sample, spacing=spacing)[0].coords
+    assert_array_equal(prop_coords, coords * np.array(spacing))
```
Comment (jni) [inline skimage/measure/tests/test_regionprops.py:343]: For all of these tests, I'd prefer if there was a separate test function for unit spacing, isotropic spacing, and anisotropic spacing, perhaps using `pytest.parametrize`? It makes it much easier to find failing tests in the future...

**Review by jni — CHANGES_REQUESTED** [review]
@tibuch awesome! Thank you for tackling this, apologies that I never got round to it! :sweat_smile: I really appreciate it. I have made a couple of suggestions, most are nitpicky and I would consider optional for approval, but the ones I really care about are adding `indices` or `coords_scaled` as a property (since coords are in many cases used for indexing so you don't want the scaling), and removing `pixel_area` as a property since it is a property of the image, not the region, and trivial to compute outside of the function.

Thank you again!

File: skimage/measure/_regionprops.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tibuch) [reply skimage/measure/_regionprops.py]: Agreed, updated accordingly.

File: skimage/measure/_regionprops.py (hunk 3 above)
Comment (tibuch) [reply skimage/measure/_regionprops.py:564]: Took me a couple hours to come up with that implementation :rofl:

File: skimage/measure/_regionprops.py (hunk 1 above)
Comment (tibuch) [reply skimage/measure/_regionprops.py:412]: Introduced `coords_scaled`.

File: skimage/measure/tests/test_regionprops.py (hunk 5 above)
Comment (tibuch) [reply skimage/measure/tests/test_regionprops.py:355]: Makes sense. 

You will also notice that not all features are tested with anisotropy, because I could not find reference numbers. Do you know of some implementations that could be used?

File: skimage/measure/_regionprops.py (hunk 4 above)
Comment (tibuch) [reply skimage/measure/_regionprops.py]: I hope I caught all of them :innocent:

File: skimage/measure/tests/test_regionprops.py (hunk 5 above)
Comment (jni) [reply skimage/measure/tests/test_regionprops.py:355]: I don't. Typically I would try to hand-check the numbers (going step by step in the implementation) and then save those as a reference. Or you can derive some from first principles, for example, drawing an ellipse with a particular orientation using `skimage.draw`, then taking away every other row of pixels and using `spacing=(2, 1)` and making sure that the results are close to the isotropic case.

**Review by jni — APPROVED** [review]
I'm approving, but will let this sit for a few days to give others a chance to review and @tibuch a chance to investigate some of the additional tests if possible. Thank you again @tibuch!

File: skimage/measure/_moments.py [hunk 6]
```diff
@@ -242,10 +246,14 @@ def moments_central(image, center=None, order=3, **kwargs):
     """
     if center is None:
         center = centroid(image)
```
[the remark this answers is not in the document]
Comment (grlee77) [reply skimage/measure/_moments.py:248]: Current tests pass with or without this, indicating we should add another test case to cover this

**Review by grlee77 — CHANGES_REQUESTED** [review]
Thanks @tibuch, this is looking pretty good. I just had some small fixes and suggestions.

**Comment by tibuch** [comment]
I implemented the requested changes.

Additionally I found that the normalized moments were wrong. Could you have a close look at this commit: [41322934dcfbb5db74b98e369d8264997465a43a](https://github.com/scikit-image/scikit-image/pull/6296/commits/41322934dcfbb5db74b98e369d8264997465a43a)

Thanks!

**Comment by jni** [comment]
Nice catch @tibuch! Looks good to me.

It looks like the tolerance on some of the tests needs to be reduced to `decimal=0`, unfortunately. I think that's fine. This is one of the failures. Since it doesn't fail in every Python version or OS, I guess it is machine-dependent. But still looks right to me.

>       assert_almost_equal(length, target_length, decimal=1)
E       AssertionError: 
E       Arrays are not almost equal to 1 decimals
E        ACTUAL: 17.955520462276805
E        DESIRED: 17.73962515156774

**Comment by tibuch** [comment]
Don't know why these are failing:
```
=========================== short test summary info ============================
FAILED scikit-image/skimage/data/tests/test_data.py::test_eagle - ValueError:...
FAILED scikit-image/skimage/io/tests/test_pil.py::test_imread_index_png_with_alpha
FAILED scikit-image/skimage/io/tests/test_pil.py::test_extreme_palette - Valu...
===== 3 failed, 8076 passed, 15 skipped, 24 warnings in 290.59s (0:04:50) ======
```

**Comment by jni** [comment]
Pillow 9.1 was released Apr 1 so I presume that is the culprit, nothing to do with your work @tibuch!

**Comment by jni** [comment]
Yeah, see #6314

**Comment by alexdesiqueira** [comment]
Thank you @tibuch! How do you like this one after the changes, @grlee77?

**Review by grlee77 — APPROVED** [review]
Thanks, @tibuch. The tests are looking good now. Thank you for fixing the normalized moments. I may take a quick look about using `pytest.parametrize` as @jni requested, but that is just a test style concern, so I wil go ahead and approve now.

**Comment by tibuch** [comment]
Hi,

I was wondering when this code will make it into an official release?

Cheers

**Comment by lagru** [comment]
Hey @tibuch. We've [just started with the preparation of the next release 0.20](https://github.com/scikit-image/scikit-image/pull/6556). We are not certain on the precise target date yet but somewhere around 2022-10-24. You can follow along our [0.20 milestone](https://github.com/scikit-image/scikit-image/milestone/22) if you want to stay informed on the progress.
